### PR TITLE
Dev private events

### PR DESCRIPTION
**Pull request by Wusinho**
The goal of this project was to create an application using Ruby on Rails for creating events including the model structure, associations, controllers, views, validations, and authentications, stylized with Bootstrap.
## Built with

- Ruby on Rails
- VS Code
- Bootstrap
- Rubocop

### CODE REVIEW

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Juan Lazo Benza** 
Good job with your first submission for this project :muscle:  :muscle:
It most of the requirements :clap: :clap: :clap:  
There are just a few details to improve before you can go ahead:

- [ ] Fix rubocop offenses
- [ ] Use Rspec + Capybara for the unit and integration tests







- [ ] User should be able to attend events
- [ ] the show page of an event should also show the list of attendants


 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

File: README.md [hunk 1]
```diff
@@ -1,24 +1,93 @@
...
+  $ npm install --global yarn
+
+- Step 2:
+  Run in the terminal:
+  $ Rails server
+
+- Step 3:
+  Open a localhost:3000 tab in your browser
+
+### How to use the aplication
```
Comment (enaburekhan) [inline README.md:52]: Correct spelling error: it should be `application`

File: app/controllers/events_controller.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+      @event = Event.new
+    end
+  end
+
+  # GET /events/1/edit
+  def edit; end
+
+  # POST /events or /events.json
+  def create
+    @event = Event.new(event_params)
```
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: uses #build method on association, i.e. current_user.events.build(event_params).

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -0,0 +1,74 @@
+class EventsController < ApplicationController
+  before_action :set_event, only: %i[show edit update destroy]
+
+  # GET /events or /events.json
+  def index
+    @events = Event.all
+  end
```
Comment (enaburekhan) [inline app/controllers/events_controller.rb]: It should list the past and upcoming events. E,g 
def index
  @past_events = Event.previous_events
  @upcoming_events = Event.upcoming_events
end

File: app/views/events/_show.html.erb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        <td></td>
+        <% end %>
+        <% if set_current_user && !set_current_user.attended_events.where(id: event.id).any? %>
+
+          <td><%= link_to 'Attend', attendances_path(event_id: event.id), method: :post, class: "btn btn-success" %></td>
+          <td></td>
+          <% else %>
+          <td><button type="button" class="btn btn-light btn-sm disabled">Already attending</button></td>
+
+        <% end %>
```
Comment (enaburekhan) [inline app/views/events/_show.html.erb]: There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. Kindly move them to a helper.

File: app/views/events/show.html.erb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      <% end %>
+   </tr>
+ 
+</ul>
+
+<div class="text-center">
+<% if set_current_user && set_current_user.id == @event.user_id %>
+<%= link_to 'Edit', edit_event_path(@event) , class: "btn btn-warning " %> 
+<% end %>
+<%= link_to 'Back', events_path , class: "btn btn-secondary mt-4" %>
```
Comment (enaburekhan) [inline app/views/events/show.html.erb]: Also, move this logic to a helper.

File: app/views/users/new.html.erb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+<%= form_with model: @user do |form| %>
+
+    <% if @user.errors.any? %>
+        <div class="alert alert-danger">
+            <% @user.errors.full_messages.each do |message| %>
+                <div><%= message %></div>
+        </div>
+    <% end %>
+    <% end %>
```
Comment (enaburekhan) [inline app/views/users/new.html.erb:12]: Also, move this logic to a helper.

File: app/views/users/sign_up.html.erb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+
+<%= form_with model: @user, url: sign_up_path do |form| %>
+
+    <% if @user.errors.any? %>
+        <div class="alert alert-danger">
+            <% @user.errors.full_messages.each do |message| %>
+                <div><%= message %></div>
+        </div>
+    <% end %>
+    <% end %>
```
Comment (enaburekhan) [inline app/views/users/sign_up.html.erb:12]: Also, move this to a helper

File: app/views/users/show.html.erb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    <tr>
+    <% if DateTime.now.utc <= event.date %>
+       <td><%= link_to  event.title, event_path(event.id) %></td>
+      <td><%= event.date %></td>
+      <td><%= @date %></td>
+    <% end %>
+     
+    </tr>
+    <% end %>
+    <% else %>
```
Comment (enaburekhan) [inline app/views/users/show.html.erb]: Also, move this logic to a helper.

File: app/views/users/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  </thead>
+  <tbody>
+  
+          <% @my_events.each do |event| %>
+    <tr>
+    <% if DateTime.now.utc > event.date %>
+       <td><%= link_to  event.title, event_path(event.id) %></td>
+      <td><%= event.date %></td>
+      <td><%= @date %></td>
+    <% end %>
```
Comment (enaburekhan) [inline app/views/users/show.html.erb]: Move this logic to a helper

**Review by enaburekhan — CHANGES_REQUESTED** [review]
# Status: Changes required:full_moon:
**Hi team!
Great job so far. Thank you for implementing some of the changes requested by the last reviewer:clap::clap:.**

**major highlights**

- The UI is awesome and logged in user can create new event
- Navigation links are present at the top of the page.
- Event's index page displays past and upcoming events list.

However, we still need to make a few changes to get the project approved:muscle:.

**To improve**

**EventsController**

- [x] `create` action

-  uses #build method on association, i.e. `current_user.events.build(event_params)` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/build_method_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). You did implement this. But the Build method is missing. Pls add it.

- [x] `index` action

- lists past events, i.e. `@past_events = Event.previous_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). This is currently missing. Please add it.

- lists ucomping events, i.e. `@upcoming_events = Event.upcoming_events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/past_upcoming_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations)). Also account for this.

- [x] There should be absolutely NO logic in views. Views are only for displaying information that became available to it from its corresponding controller not generating it. As already highlighted in the PR below, Move all logics in view to a helper. Check this [article](http://codefol.io/posts/Where-Do-I-Put-My-Code/) and this [article](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers) on how to use a helper in rails. Also, see [screen-shot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/section/no_logic_in_helpers_requirement.png) from pathwright.

- [x] Remove commented-out code from events_controller.rb. and attendances_controller.rb. They make the codebase to be untidy and the codes difficult to read and understand. 

Make the above-required changes and when you are ready, request another review.
Kindly drop your questions :point_down: and ping me on slack @ Eric Enaburekhan to respond.
Happy coding:blush:
Eric Enaburekhan

File: app/views/users/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+  </thead>
+  <tbody>
+  
+          <% @my_events.each do |event| %>
+    <tr>
+    <% if DateTime.now.utc <= event.date %>
+       <td><%= link_to  event.title, event_path(event.id) %></td>
+      <td><%= event.date %></td>
+      <td><%= @date %></td>
+    <% end %>
```
Comment (whiz25) [inline app/views/users/show.html.erb]: - Move this logic to a helper

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Changes Required 🙏🏿 
Amazing work. You made most of the changes requested. Well done!
You, however, missed a requested change, there's still some logic in views. I left you a comment with a suggestion.
Make the change and submit it for another review.


### Martin Afani
#### Slack: whiz25

**Review by NtwaliHeritier — APPROVED** [review]
### **STATUS: APPROVED✔✔✔✔**
Great implementation of the project👏👏
You can merge👌👌👍👍

Happy coding💻💻💻

**Pull request merged**